cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go

```diff
@@ -822,70 +826,106 @@ func (s *Server) CreateInstalledPackage(ctx context.Context, request *corev1.Cre
+	// We may want to include the AvailablePackageRef in the request, given
+	// that it can be ambiguous, but we dont yet have a UI that allows the
+	// user to select which chart to use, so until then.
+	detailResponse, err := s.GetInstalledPackageDetail(ctx, &corev1.GetInstalledPackageDetailRequest{
+		InstalledPackageRef: installedRef,
+	})
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Unable to create Helm action config: %v", err)
+		return nil, err
+	}
+
+	availablePkgRef := detailResponse.GetInstalledPackageDetail().GetAvailablePackageRef()
+	if availablePkgRef == nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "Unable to find the available package used to deploy %q in the namespace %q.", releaseName, installedRef.GetContext().GetNamespace())
 	}
 
-	// We currently get app repositories on the kubeapps cluster only.
 	typedClient, _, err := s.GetClients(ctx, s.globalPackagingCluster)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "Unable to create kubernetes clientset: %v", err)
 	}
-	registrySecrets, err := chart.RegistrySecretsPerDomain(ctx, appRepo.Spec.DockerRegistrySecrets, appRepo.Namespace, typedClient)
+	chartID := availablePkgRef.GetIdentifier()
+	repoName, chartName, err := splitChartIdentifier(chartID)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Unable to fetch registry secrets from the namespace %q: %v", appRepo.Namespace, err)
+		return nil, err
+	}
+	chartDetails := &chartutils.Details{
+		AppRepositoryResourceName:      repoName,
+		AppRepositoryResourceNamespace: availablePkgRef.GetContext().GetNamespace(),
+		ChartName:                      chartName,
+		Version:                        request.GetPkgVersionReference().GetVersion(),
 	}
+	ch, registrySecrets, err := s.fetchChartWithRegistrySecrets(ctx, chartDetails, typedClient)
 
-	release, err := agent.CreateRelease(actionConfig, request.GetName(), namespace, request.GetValues(), ch, registrySecrets)
+	// Create an action config for the installed pkg context.
+	actionConfig, err := s.actionConfigGetter(ctx, installedRef.GetContext())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Unable to create helm release %q in the namespace %q: %v", request.GetName(), appRepo.Namespace, err)
+		return nil, status.Errorf(codes.Internal, "Unable to create Helm action config: %v", err)
 	}
 
-	return &corev1.CreateInstalledPackageResponse{
+	release, err := agent.UpgradeRelease(actionConfig, releaseName, request.GetValues(), ch, registrySecrets)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to upgrade helm release %q in the namespace %q: %v", releaseName, installedRef.GetContext().GetNamespace(), err)
+	}
+
+	return &corev1.UpdateInstalledPackageResponse{
 		InstalledPackageRef: &corev1.InstalledPackageReference{
 			Context: &corev1.Context{
-				Cluster:   cluster,
+				Cluster:   installedRef.GetContext().GetCluster(),
 				Namespace: release.Namespace,
```

Comment:
Doubt: shouldn't cluster and namespace both be extracted from the `installedRef` ? Well, I mean, `release.Namespace` should match `installedRef.GetContext().GetNamespace()` shouldn't it?

Reply:
Yes, the namespaces should definitely match, which is why, in case of any error in logic, I go with the release namespace. But since the cluster isn't part of the release so I have no option but to use the requested one (which is what is passed to the helm action config). That said, I should be similarly ensuring that cluster is set to the default cluster if it was blank in the request here... I'll add a test and fix that too.